src/CPULSSolver.cpp

```diff
@@ -145,6 +148,150 @@ void CPULSSolver::initializeFSRs() {
 }
 
 
+/**
+ * @brief Initializes the arrays for the fixed source and its moments
+ */
+void CPULSSolver::initializeFixedSources() {
+
+  CPUSolver::initializeFixedSources();
+  log_printf(NORMAL, "Initializing linear fixed sources...");
+
+  /* Allocate the fixed sources array if not yet allocated */
+  if (_fixed_sources_xyz.empty()) {
+    long size = _num_FSRs * _num_groups;
+    _fixed_sources_xyz.resize(size);
+    for (long i=0; i<size; i++)
+      _fixed_sources_xyz.at(i).resize(3, 0);
+  }
+
+  /* Fill the fixed source moments with the user defined values */
+  Cell* fsr_cell;
+  long cell_id, fsr_id;
+  int group;
+  double source_x, source_y, source_z;
+  std::map< std::pair<Cell*, int>, std::vector<double> >::iterator cell_iter;
+  std::map< std::pair<int, int>, std::vector<double> >::iterator fsr_iter;
+
+  /* Fixed sources assigned by Cell */
+  for (cell_iter = _fix_src_xyz_cell_map.begin();
+       cell_iter != _fix_src_xyz_cell_map.end(); ++cell_iter) {
+
+    /* Get the Cell with an assigned fixed source */
+    cell_id = (cell_iter->first).first->getId();
+    group = cell_iter->first.second;
+    source_x = cell_iter->second[0];
+    source_y = cell_iter->second[1];
+    source_z = cell_iter->second[2];
+
+    /* Search for this Cell in all FSRs */
+    for (long r=0; r < _num_FSRs; r++) {
+      fsr_cell = _geometry->findCellContainingFSR(r);
```

Comment:
Can we have a more efficient way to map the cell_id with fsr_id?

Reply:
Ok I'll just precompute a map from fsr to cells

Reply:
I added a routine in geometry.cpp to compute this, since it's more in that file's scope than the solver's. In the longer run, we may want to add that map to the geometry class attribute, if there is more use for it.